src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt

```diff
@@ -281,10 +270,18 @@ class ReferralService(
       )
     }
 
-    update.serviceCategoryIds?.let {
-      referral.selectedServiceCategories = serviceCategoryRepository.findByIdIn(it)
+    update.serviceCategoryIds?.let { serviceCategoryIds ->
+      referral.complexityLevelIds = referral.complexityLevelIds?.filter { complexityLevel -> serviceCategoryIds.contains(complexityLevel.key) }?.toMutableMap()
+      referral.selectedDesiredOutcomes = referral.selectedDesiredOutcomes?.filter { desiredOutcome -> serviceCategoryIds.contains(desiredOutcome.serviceCategoryId) }?.toMutableList()
     }
 
+    // Need to save and flush the entity before removing selected service categories due to constraint violations being thrown
+    referralRepository.save(referral)
+    referralRepository.flush()
+
+    update.serviceCategoryIds?.let { serviceCategoryIds ->
+      referral.selectedServiceCategories = serviceCategoryRepository.findByIdIn(serviceCategoryIds)
+    }
```

Comment:
Need to save, flush and save to prevent the entitymanager attempting to delete the selected service categories first before the selected complexity level and desired outcomes. This would otherwise throw a constraint violation.

(This occurs even if you order sequence of statements in the correct order)